### ISSUE TITLE
index collapse when training LAQ from scratch

### ISSUE THREAD

**Issue by giuliano-97** [LatentActionPretraining/LAPA#17]
Hi, thank you very much for releasing the code for this very cool work.

I have tried to train the LAQ on SSV2 from scratch with the exact same script that you provided. Unfortunately, I've observed that index collapse inevitably occurs after a few hundreds of steps. When the codebook replacement stops triggering (after 5k steps), the model ends up using only one code. So, I tried to initialize with the OpenX weights that you recently released and the training was significantly more stable, with no mode collapse even after 25k steps with a bs of 64- likely because the codes in the pre-trained model are already "well spread out". 

After that, I tried to do the same thing, train the LAQ from scratch, on one of the Open-X datasets, specifically the BridgeV2 dataset, and unfortunately I observed the same result. Since then, I've made some progress, but I'm still struggling with instability issues in the codebook training.

So yeah, I was wondering if you could share some tips and / or additional info on how you trained the LAQ from scratch on OpenX. 

Thank you in advance!

**Comment by seonghyeonye** [comment]
Hi, we found that there was a minor bug which happened during merging multiple branches. 
Could you pull the most recent code and try it again?
Sorry for the inconvenience.

**Comment by giuliano-97** [comment]
@seonghyeonye I've applied the changes you made [in the last commit on master](https://github.com/LatentActionPretraining/LAPA/commit/c45efa264f00ef0e26493156a16d008095dbff56) to local fork and it does seem to help, the codebook hasn't collapsed after 2k steps, which is what I always observed before. That being said, are you sure that this change here https://github.com/LatentActionPretraining/LAPA/commit/c45efa264f00ef0e26493156a16d008095dbff56#diff-0a972b3fdd43bffc90de3eeaf05a64805f9c99e88109d87eecba35859fd79c00R251 is correct?
```diff
-            self.vq.replace_unused_codebooks(224*4)  
+           self.vq.replace_unused_codebooks(tokens.shape[0])
```
in this way, you are passing the (constant) batch size whereas NSVQ says you should pass the number of batches that have been observed since the last update

**Comment by seonghyeonye** [comment]
Yes, we are passing number of batches which is set constant for our runs. If collapse happens, you can try out tweaking hyperparameters such as `discarding_threshold` or the frequency of codebook replacement for nsvq.

**Comment by WeishengDAI** [comment]
> Hi, thank you very much for releasing the code for this very cool work.
> 
> I have tried to train the LAQ on SSV2 from scratch with the exact same script that you provided. Unfortunately, I've observed that index collapse inevitably occurs after a few hundreds of steps. When the codebook replacement stops triggering (after 5k steps), the model ends up using only one code. So, I tried to initialize with the OpenX weights that you recently released and the training was significantly more stable, with no mode collapse even after 25k steps with a bs of 64- likely because the codes in the pre-trained model are already "well spread out".
> 
> After that, I tried to do the same thing, train the LAQ from scratch, on one of the Open-X datasets, specifically the BridgeV2 dataset, and unfortunately I observed the same result. Since then, I've made some progress, but I'm still struggling with instability issues in the codebook training.
> 
> So yeah, I was wondering if you could share some tips and / or additional info on how you trained the LAQ from scratch on OpenX.
> 
> Thank you in advance!

@giuliano-97  Hi, I’m also training LAQ from scratch on SSV2 using exactly the same script, and I ran into the same issue. My codebook vocab size is 8, but I found that only one code is being used. I was wondering if you managed to solve this problem later. I’d really appreciate any help—thank you!
